### PR TITLE
API for shared services

### PR DESCRIPTION
**Pull request by tanya-borisova**
# PR for issue

Closes issue #1239 

## What is being addressed

Add API for shared services: creation, deletion, update, list, and operations

## How is this addressed

- [x] Models and repository for shared services, that inherits from base resource.py repository
- [x] Tests for shared services repository
- [x] API endpoints for shared services and shared service operations - same as exist for workspace operations
- [x] Tests for API endpoints for shared services
- [x] Code reuse with workspace services (some methods are extracted into `api_app/routes/resource_helpers.py` -- I will leave comments where code has been moved 
- [x] Add some fixes for `firewall` shared service that were needed to achieve a successful deploy (mostly just remove `azure_location` parameter as it isn't necessary) 

## How is it tested

I tested that locally I can build a firewall bundle and then successfully deploy it via the API (screenshot is logs of resource processor). No code changes in resource processor were necessary

MicrosoftTeams-image (1)

### CODE REVIEW

File: api_app/api/routes/resource_helpers.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+from db.repositories.operations import OperationRepository
+from db.repositories.resource_templates import ResourceTemplateRepository
+from models.domain.resource import ResourceType, Resource
+from models.domain.operation import Operation
+from resources import strings
+from service_bus.resource_request_sender import send_resource_request_message, RequestAction
+from services.authentication import get_access_service
+
+
+async def save_and_deploy_resource(resource, resource_repo, operations_repo) -> Operation:
```
Comment (tanya-borisova) [inline api_app/api/routes/resource_helpers.py:16]: Moved from https://github.com/microsoft/AzureTRE/pull/1531/files#diff-6404c8502599a653702ae036aacc2c90444c07dbd190c6be59f69e4f1098d990

File: api_app/tests_ma/test_api/test_routes/test_resource_helpers.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        action="install",
+        message="test",
+        Status=Status.Deployed,
+        createdWhen=1642611942.423857,
+        updatedWhen=1642611942.423857
+    )
+    return operation
+
+
+class TestHelpers:
```
Comment (tanya-borisova) [inline api_app/tests_ma/test_api/test_routes/test_resource_helpers.py:65]: Moved from https://github.com/microsoft/AzureTRE/pull/1531/files#diff-a55f897ac7899c23870f99a691befe5deef28c53f5506598c7f2116c8a4f8fe9

File: api_app/resources/strings.py [hunk 3]
```diff
@@ -38,6 +38,13 @@
 API_GET_SHARED_SERVICE_TEMPLATES = "Get shared service templates"
 API_GET_SHARED_SERVICE_TEMPLATE_BY_NAME = "Get shared service template by name"
 
+API_GET_ALL_SHARED_SERVICES = "Get all shared services for shared"
```
Comment (damoodamoo) [inline api_app/resources/strings.py:41]:
```suggestion
API_GET_ALL_SHARED_SERVICES = "Get all shared services"
```

File: api_app/api/routes/shared_services.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    operation = await save_and_deploy_resource(shared_service, shared_services_repo, operations_repo)
+    response.headers["Location"] = construct_location_header(operation)
+
+    return OperationInResponse(operation=operation)
+
+
+@shared_services_router.patch("/shared-services/{shared_service_id}", response_model=SharedServiceInResponse, name=strings.API_UPDATE_SHARED_SERVICE, dependencies=[Depends(get_current_admin_user), Depends(get_shared_service_by_id_from_path)])
+async def patch_shared_service(shared_service_patch: ResourcePatch, shared_service_repo=Depends(get_repository(SharedServiceRepository)), shared_service=Depends(get_shared_service_by_id_from_path), resource_template_repo=Depends(get_repository(ResourceTemplateRepository)), etag: str = Header(None)) -> SharedServiceInResponse:
+    check_for_etag(etag)
```
Comment (damoodamoo) [inline api_app/api/routes/shared_services.py:52]: my 'Resource updates' merge has just gone in underneath you - I wonder if now would be a good chance to copy that logic to handle the updates of shared services in the `PATCH` (see workspace patch route in latest main) ? 'should' only be a few lines change in this method...

Comment (tanya-borisova) [reply api_app/api/routes/shared_services.py:52]: That sounds like a good idea, will look into it now...

**Review by damoodamoo — APPROVED** [review]

**Comment by damoodamoo** [comment]
cracking work - LGTM :)

**Pull request merged**